### PR TITLE
User/ntalokar/worker classes

### PR DESCRIPTION
**Pull request by nidhitalokar**

### CODE REVIEW

File: app/src/main/java/com/example/vaccinespotter/MainActivity.java [hunk 1]
```diff
@@ -1,60 +1,42 @@
...
 
-import retrofit2.Call;
-import retrofit2.Callback;
-import retrofit2.Response;
-import retrofit2.Retrofit;
-import retrofit2.converter.gson.GsonConverterFactory;
+import java.util.concurrent.TimeUnit;
 
 public class MainActivity extends AppCompatActivity {
+    public static Long REPEAT_INTERVAL = 15L;
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/MainActivity.java]: Please make these final

File: app/src/main/java/com/example/vaccinespotter/MainActivity.java [hunk 2]
```diff
@@ -1,60 +1,42 @@
...
+    public static TimeUnit REPEAT_INTERVAL_UNIT = TimeUnit.MINUTES;
+    public static String workerTag = "VaccineWorkerTag";
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/MainActivity.java]: Please follow naming convention `WORKER_TAG`

File: app/src/main/java/com/example/vaccinespotter/RetrofitManager.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            .addConverterFactory(GsonConverterFactory.create())
+            .build();
+
+    private final VaccineSlots vaccineSlots = retrofit.create(VaccineSlots.class);
+
+    public void QueryCowin(Context context, String date) {
+
+        VaccineNotificationManager notificationManager = new VaccineNotificationManager(context);
+        notificationManager.registerNotifications();
+        int districtId = 312;
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/RetrofitManager.java]: Please extract this as a const member

File: app/src/main/java/com/example/vaccinespotter/RetrofitManager.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        Call<Centers> call = vaccineSlots.getCenters(districtId, date);
+
+        call.enqueue(new Callback<Centers>() {
+            @Override
+            public void onResponse(Call<Centers> call, Response<Centers> response) {
+                if (response.isSuccessful()) {
+                    // This contains the response.
+                    Centers centers = response.body();
+                    Toast.makeText(context, "Success", Toast.LENGTH_SHORT).show();
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/RetrofitManager.java]: Is this required now?

File: app/src/main/java/com/example/vaccinespotter/RetrofitManager.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                    Center center = centers.getCenters().get(0);
+                    notificationManager.notifyUser(new NotificationModel((CenterBase) center, center.getSessions()[0]));
+                    notificationManager.notifyUser(new NotificationModel((CenterBase) center, center.getSessions()[0]));
+
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/RetrofitManager.java]: This needs to be fixed, it was just for example.

File: app/src/main/java/com/example/vaccinespotter/RetrofitManager.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                    notificationManager.notifyUser(new NotificationModel((CenterBase) center, center.getSessions()[0]));
+                }
+            }
+
+            @Override
+            public void onFailure(Call<Centers> call, Throwable t) {
+
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/RetrofitManager.java]: Maybe can do something here to notify the user?

File: app/src/main/java/com/example/vaccinespotter/RetrofitManager.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    private final VaccineSlots vaccineSlots = retrofit.create(VaccineSlots.class);
+
+    public void QueryCowin(Context context, String date) {
+
+        VaccineNotificationManager notificationManager = new VaccineNotificationManager(context);
+        notificationManager.registerNotifications();
+        int districtId = 312;
+        Call<Centers> call = vaccineSlots.getCenters(districtId, date);
+
+        call.enqueue(new Callback<Centers>() {
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/RetrofitManager.java]: Make this synchronous

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+public class BackgroundWorker extends Worker {
+
+    public BackgroundWorker(
+            @NonNull @NotNull Context context,
+            @NonNull @NotNull WorkerParameters workerParams) {
+        super(context, workerParams);
+    }
+
+    @NonNull
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java:29]: Not sure if these annotations are required. Please remove these.

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import org.jetbrains.annotations.NotNull;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Locale;
+
+public class BackgroundWorker extends Worker {
+
+    public BackgroundWorker(
+            @NonNull @NotNull Context context,
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: Not sure if these annotations are required. Please remove these.

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+            @NonNull @NotNull WorkerParameters workerParams) {
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: Not sure if these annotations are required. Please remove these.

**Review by drhousemd — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    @NonNull
+    @org.jetbrains.annotations.NotNull
+    @Override
+    public Result doWork() {
+        RetrofitManager retrofitManager = new RetrofitManager();
+
+        DateFormat dateFormat = new SimpleDateFormat("dd-MM-yyyy", Locale.US);
+        Calendar calendar = Calendar.getInstance();
+        String date = dateFormat.format(calendar.getTime());
+        for(int cycles = 0; cycles <= 5 ;cycles++)
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: whitespace after for

Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: Extract this 5 as a constant.

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            retrofitManager.QueryCowin(getApplicationContext(), date);
+            calendar.add(Calendar.DATE, 7);
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: Extract this 7 as a constant.

**Review by drhousemd — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/vaccinespotter/RetrofitManager.java (hunk 4 above)
Comment (nidhitalokar) [reply app/src/main/java/com/example/vaccinespotter/RetrofitManager.java]: Changed it to sync.

File: app/src/main/java/com/example/vaccinespotter/RetrofitManager.java (hunk 5 above)
Comment (nidhitalokar) [reply app/src/main/java/com/example/vaccinespotter/RetrofitManager.java]: Done

File: app/src/main/java/com/example/vaccinespotter/RetrofitManager.java (hunk 6 above)
Comment (nidhitalokar) [reply app/src/main/java/com/example/vaccinespotter/RetrofitManager.java]: Introduced notifications for failure.

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+import com.example.vaccinespotter.notifications.VaccineNotificationManager;
+
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Locale;
+
+public class BackgroundWorker extends Worker {
+    private static final String TAG = "BackgroundWorker";
+    private static final int numberOfCycles = 5;
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: static members naming convention https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html

If your variable stores a constant value, such as static final int NUM_GEARS = 6, the convention changes slightly, capitalizing every letter and separating subsequent words with the underscore character. By convention, the underscore character is never used elsewhere.

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    private static final int numberOfDaysInCycle = 7;
+    private VaccineNotificationManager notificationManager = null;
+    private String queryCowinFailed = "Querying the CoWin website failed";
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: these can be final and static

File: app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.List;
+import retrofit2.Call;
+import retrofit2.Response;
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+public class RetrofitManager {
+    private final String TAG = "RetrofitManager";
+    private static final int age18 = 18;
+    private static final int districtId = 312;
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java]: If your variable stores a constant value, such as static final int NUM_GEARS = 6, the convention changes slightly, capitalizing every letter and separating subsequent words with the underscore character. By convention, the underscore character is never used elsewhere.

File: app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+    private boolean isVaccineAvailable(Session session) {
+        return (session.getAvailable_capacity() != 0);
+    }
+
+    private boolean checkFor18YearsPlus(Session session) {
+        return (session.getMin_age_limit() == age18);
+    }
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java:43]: These methods belong in the Sessions class

File: app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+            Log.e(TAG, "Error in querying CoWin: " + ex);
+            return false;
+        }
+    }
+
+    private boolean isVaccineAvailable(Session session) {
+        return (session.getAvailable_capacity() != 0);
+    }
+
+    private boolean checkFor18YearsPlus(Session session) {
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java]: better name `isFor18YearsPlus()`

File: app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    private static final int age18 = 18;
+    private static final int districtId = 312;
+    private final Retrofit retrofit = new Retrofit.Builder()
+            .baseUrl("https://cdn-api.co-vin.in")
+            .addConverterFactory(GsonConverterFactory.create())
+            .build();
+
+    private final VaccineSlots vaccineSlots = retrofit.create(VaccineSlots.class);
+
+    public boolean queryCowin(VaccineNotificationManager notificationManager, Context context, String date) {
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java]: Better to not have this dependency of NotificationManager here and return the list of valid NotificationModel.
You can then notify user from the worker itself.

File: app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+        try {
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/RetrofitManager.java]: Can remove the try catch from here and specify what exception this method throws.

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Locale;
+
+public class BackgroundWorker extends Worker {
+    private static final String TAG = "BackgroundWorker";
+    private static final int numberOfCycles = 5;
+    private static final int numberOfDaysInCycle = 7;
+    private VaccineNotificationManager notificationManager = null;
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: `mNotificationManager`

File: app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    public Result doWork() {
+        try {
+            notificationManager = new VaccineNotificationManager(getApplicationContext());
+            notificationManager.registerNotifications();
+            RetrofitManager retrofitManager = new RetrofitManager();
+
+            DateFormat dateFormat = new SimpleDateFormat("dd-MM-yyyy", Locale.US);
+            Calendar calendar = Calendar.getInstance();
+            String date = dateFormat.format(calendar.getTime());
+            for (int cycles = 0; cycles <= numberOfCycles; cycles++)
```
Comment (drhousemd) [inline app/src/main/java/com/example/vaccinespotter/worker/BackgroundWorker.java]: remove <=

**Review by drhousemd — APPROVED** [review]

**Pull request merged**